File: 54. 螺旋矩阵.scala
// 给定一个包含 m x n 个元素的矩阵（m 行, n 列），请按照顺时针螺旋顺序，返回矩阵中的所有元素。

// 示例 1:
// 输入:
// [
//  [ 1, 2, 3 ],
//  [ 4, 5, 6 ],
//  [ 7, 8, 9 ]
// ]
// 输出: [1,2,3,6,9,8,7,4,5]

// 示例 2:
// 输入:
// [
//   [1, 2, 3, 4],
//   [5, 6, 7, 8],
//   [9,10,11,12]
// ]
// 输出: [1,2,3,4,8,12,11,10,9,5,6,7]

import scala.collection.mutable.ListBuffer

object Solution {
  def spiralOrder(matrix: Array[Array[Int]]): List[Int] = {
    if (matrix.length == 0) return Nil
    val (m, n) = (matrix.length, matrix(0).length)
    val res = ListBuffer[Int]()
    0 to ((n - 1) / 2 min (m - 1) / 2 min m min n) foreach { i =>
      val w = n - 2 * i
      val h = m - 2 * i
      if (h == 1) i until i + w foreach (res += matrix(i)(_))
      else if (w == 1) i until i + h foreach (res += matrix(_)(i))
      else {
        i until i + w foreach (res += matrix(i)(_))
        i + 1 until i + h - 1 foreach (res += matrix(_)(i + w - 1))
        i + w - 1 until i - 1 by -1 foreach (res += matrix(i + h - 1)(_))
        i + h - 2 until i by -1 foreach (res += matrix(_)(i))
      }
    }
    res.toList
  }
}

File: 54. 螺旋矩阵.py
# 给你一个 m 行 n 列的矩阵   matrix ，请按照 顺时针螺旋顺序 ，返回矩阵中的所有元素。
#
# 示例 1：
# 输入：matrix = [[1,2,3],[4,5,6],[7,8,9]]
# 输出：[1,2,3,6,9,8,7,4,5]
#
# 示例 2：
# 输入：matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
# 输出：[1,2,3,4,8,12,11,10,9,5,6,7]
#
# 提示：
# m == matrix.length
# n == matrix[i].length
# 1 <= m, n <= 10
# -100 <= matrix[i][j] <= 100

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []

        def helper(ai: int, bi: int, aj: int, bj: int):
            for j in range(aj, bj + 1):
                ans.append(matrix[ai][j])
            for i in range(ai + 1, bi):
                ans.append(matrix[i][bj])
            # 防止重复
            if ai != bi:
                for j in range(bj, aj - 1, -1):
                    ans.append(matrix[bi][j])
            if aj != bj:
                for i in range(bi - 1, ai, -1):
                    ans.append(matrix[i][aj])

        ai, bi, aj, bj = 0, len(matrix) - 1, 0, len(matrix[0]) - 1
        while ai <= bi and aj <= bj:
            helper(ai, bi, aj, bj)
            ai += 1
            aj += 1
            bi -= 1
            bj -= 1
        return ans
